src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -180,6 +187,15 @@ private async Task<(Document document, SemanticModel? semanticModel)> GetDocumen
 
             return MergeAndPruneCompletionLists(allContexts, defaultItemSpan, options, isExclusive: false);
 
+            AsyncLazy<SyntaxContext?> GetLazySyntaxContextWithSpeculativeModel(Document document, int position)
+            {
+                if (!relatedDocumentIds.Contains(document.Id) || caretPosition != position)
+                    throw new ArgumentException("Don't support getting SyntaxContext for document unrelated to the original document or for a different position");
+
+                return syntaxContextCache.GetOrAdd(document, d => AsyncLazy.Create(c
+                     => CompletionHelper.CreateSyntaxContextWithExistingSpeculativeModelAsync(document, position, c), cacheResult: true));
```

Comment:
The use of `position` here will cause a capture to be created (allocation) even in cases where the item is already in the dictionary. In such a case, it's better to use `TryGetValue` first, and then create private helper method `GetLazySyntaxContextWithSpeculativeModelSlow` that calls `GetOrAdd`.

Reply:
💡 It would be best to place the `TryGetValue` even before the `relatedDocumentIds` check (but after the `position` check). This would fully optimize the path once the model exists.